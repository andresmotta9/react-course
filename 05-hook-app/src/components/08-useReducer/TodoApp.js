import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import './style.css';  
import { TodoList } from './components/TodoList';
import { TodoAdd } from './components/TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) )
    }, [todos]);

    const handleDelete = (todoId) => {
        
        const action = { 
            type: 'delete',
            payload: todoId,
        }

        dispatch(action);
    }
    const handleToggle = ( todoId ) => {
        dispatch({
            type: 'toggle',
            payload: todoId,
        })
    }

    const hanldeAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo,
        });
    }

    return (
        <div>
            <h1>TodoApp ( {todos.length} )</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <TodoList 
                        todos={todos}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                </div>
                <div className="col-5">
                   <TodoAdd hanldeAddTodo={hanldeAddTodo}/>
                </div>
            </div>

        </div>
    )
}
