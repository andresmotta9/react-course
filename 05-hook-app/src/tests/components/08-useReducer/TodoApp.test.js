import React from 'react';
import { shallow } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Tests on <TodoApp />', () => {
    
    const wrapper = shallow( <TodoApp />);

    test('should show correctly', () => {

        expect( wrapper ).toMatchSnapshot();

    });

    test('should delete a TODO', () => {
        
        wrapper.find('TodoAdd').prop('hanldeAddTodo')( demoTodos[0] );
        wrapper.find('TodoList').prop('handleDelete')( demoTodos[0].id );

        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp ( 0 )');
    });
    
    
})  
