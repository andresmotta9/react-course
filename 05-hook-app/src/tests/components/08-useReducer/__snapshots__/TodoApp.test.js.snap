// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TodoApp /> should show correctly 1`] = `
<div>
  <h1>
    TodoApp ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        hanldeAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
